src/pool/srv_pool.c

```diff
@@ -3917,48 +3921,100 @@ replace_failed_replicas(struct pool_svc *svc, struct pool_map *map)
+ * \param[in]	tgt_addrs	optional target addresses (ignored if \a tgts is
+ *				nonempty; requires inval_tgt_addrs)
+ * \param[out]	hint		optional leadership hint
+ * \param[out]	p_updated	optional info on if pool map has been updated
+ * \param[out]	map_version_p	pool map version
+ * \param[out]	tgt_map_ver	pool map version for the last target change
+ *				(instead of a node change, for example) made by
+ *				this update, or 0 if none has been made (see
+ *				ds_pool_map_tgts_update)
+ * \param[out]	inval_tgt_addrs	optional invalid target addresses (ignored if
+ *				\a tgts is nonempty; if specified, must be
+ *				initialized to empty and freed by the caller)
+ */
 static int
-ds_pool_update_internal(uuid_t pool_uuid, struct pool_target_id_list *tgts,
-			unsigned int opc, struct rsvc_hint *hint,
-			bool *p_updated, bool evict_rank,
-			uint32_t *map_version_p, uint32_t *tgt_map_ver)
+pool_svc_update_map_internal(struct pool_svc *svc, unsigned int opc,
+			     bool exclude_rank,
+			     struct pool_target_id_list *tgts,
+			     struct pool_target_addr_list *tgt_addrs,
+			     struct rsvc_hint *hint, bool *p_updated,
+			     uint32_t *map_version_p, uint32_t *tgt_map_ver,
+			     struct pool_target_addr_list *inval_tgt_addrs)
 {
-	struct pool_svc	       *svc;
 	struct rdb_tx		tx;
-	struct pool_map	       *map = NULL;
+	struct pool_map	       *map;
 	uint32_t		map_version_before;
-	uint32_t		map_version = 0;
-	struct pool_buf	       *map_buf = NULL;
+	uint32_t		map_version;
+	struct pool_buf	       *map_buf;
 	bool			updated = false;
 	int			rc;
 
-	rc = pool_svc_lookup_leader(pool_uuid, &svc, hint);
-	if (rc != 0)
-		D_GOTO(out, rc);
+	D_DEBUG(DB_MD,
+		DF_UUID": opc=%u exclude_rank=%d ntgts=%d ntgt_addrs=%d\n",
+		DP_UUID(svc->ps_uuid), opc, exclude_rank, tgts->pti_number,
+		tgt_addrs == NULL ? 0 : tgt_addrs->pta_number);
 
 	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
 	if (rc != 0)
-		D_GOTO(out_svc, rc);
+		goto out;
 	ABT_rwlock_wrlock(svc->ps_lock);
 
 	/* Create a temporary pool map based on the last committed version. */
 	rc = read_map(&tx, &svc->ps_root, &map);
 	if (rc != 0)
-		D_GOTO(out_map, rc);
+		goto out_lock;
+
+	/*
+	 * If an empty target ID list is provided, convert from target
+	 * addresses.
+	 */
+	if (tgts->pti_number == 0) {
+		D_ASSERT(tgts->pti_ids == NULL);
+		D_ASSERT(tgt_addrs != NULL);
+		D_ASSERT(inval_tgt_addrs != NULL);
+		rc = pool_find_all_targets_by_addr(map, tgt_addrs, tgts,
+						   inval_tgt_addrs);
+		if (rc != 0)
+			goto out_map;
+		if (inval_tgt_addrs->pta_number > 0) {
+			/*
+			 * If any invalid ranks/targets were specified here,
+			 * abort the entire request. This will mean the
+			 * operator needs to resubmit the request with
+			 * corrected arguments, which will be easier without
+			 * trying to figure out which arguments were accepted &
+			 * started processing already.
+			 */
+			rc = -DER_NONEXIST;
```

Comment:
this probably worth a DB_MD message.  btw: will -DER_NONEXIST be ignored by dmg pool cmd ?  i.e. 

dmg pool exclude tgts will return 0 or ENONEXIST if there are some targets which do not exist?

Reply:
Done.